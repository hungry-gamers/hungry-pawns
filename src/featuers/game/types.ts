export type PawnSize = 'small' | 'medium' | 'big'

export interface Pawn {
  size: PawnSize
  playerId: string
}

export type Move = {
  size: PawnSize
  mode: 'pawn' | 'shield'
}

export type Location = { rowIndex: number; columnIndex: number }

export type ApplyShieldPayload = Location

export type SpecialMoveName = 'shield'

export type Sequence = `${string}/${number}-${number}/${number}-${number}/${number}-${number}`

export type Cell = { pawn: Pawn | null; shield: { activeInTurn: number; appliedBy: string } }
export type Board = Cell[][]
export type Player = {
  pawns: Record<PawnSize, number>
  capturedPawnsCounter: number
  arePawnsLocked: boolean
  specialMoves: SpecialMoveName[]
}

export type PlayerPayload = { id: string; pawns: Player['pawns'] }

export type PutPawnPayload = { pawnSize: PawnSize } & Location

export type Game = {
  status: 'not-initialized' | 'pregame' | 'in-progress' | 'finished'
  board: Board
  currentPlayerId: string
  turns: Record<number, { playerId: string; move: PutPawnPayload | ApplyShieldPayload }>
  winner: string | undefined
  allowedPawns: PawnSize[]
  sequences: Sequence[]
}
