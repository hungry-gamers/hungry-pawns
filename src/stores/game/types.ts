export type PawnSize = 'small' | 'medium' | 'big'

export interface Pawn {
  size: PawnSize
  playerId: string
}

export type Cell = Pawn | null
export type Board = (Cell | null)[][]
export type Player = {
  id: string
  pawns: {
    small: number
    medium: number
    big: number
  }
}

export type PutPawnPayload = { pawnSize: PawnSize; rowIndex: number; columnIndex: number }

export type Game = {
  status: 'not-initialized' | 'pregame' | 'in-progress' | 'finished'
  board: Board
  currentPlayerId: string
  pawns: Record<string, Record<PawnSize, number>>
  pawnsLockedBy: string[]
  turns: Record<number, { playerId: string; move: PutPawnPayload }>
  potentialWinner: string | undefined
  eatenPawnsCounter: Record<string, number>
}

type WinCondition = 'captured-line' | 'pawns-eaten' | 'balance' | 'dominance'

export type WinData = { winnerId: string; winCondition: WinCondition } | { winnerId: undefined }
