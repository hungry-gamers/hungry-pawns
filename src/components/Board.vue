<script setup lang="ts"></script>

<template>
  <div id="board">
    <div class="cell"></div>
  </div>
</template>

<style scoped>
#board {
  display: grid;
  grid-template-columns: repeat(3, 100px);
  grid-template-rows: repeat(3, 100px);
  gap: 5px;
}

.cell {
  width: 100px;
  height: 100px;
  background-color: white;
  border: 2px solid black;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 2rem;
  cursor: pointer;
}
</style>
