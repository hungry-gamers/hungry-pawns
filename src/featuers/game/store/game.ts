import { reactive } from 'vue'
import { defineStore } from 'pinia'
import * as GameT from '@/featuers/game/types.ts'
import * as PlayersT from '@/featuers/players/types'
import * as GameService from '@/featuers/game/services/game.service.ts'
import { usePlayersStore } from '@/featuers/players/store/players.ts'
import { SpecialEffectsSequences } from '@/featuers/game/services/game.service.ts'

export const useGameStore = defineStore('game', () => {
  const playersStore = usePlayersStore()

  const state = reactive<GameT.Game>({
    status: 'not-initialized',
    board: GameService.createBoard(),
    currentPlayerId: '',
    turns: {},
    winner: undefined,
    allowedPawns: ['small'],
    sequences: [],
  })

  const initiateGame = (players: PlayersT.PlayerPayload[]) => {
    state.status = 'pregame'
    state.board = GameService.createBoard()
    state.currentPlayerId = players[0].id
    state.allowedPawns = ['small']
    state.sequences = []
    state.winner = undefined

    playersStore.createPlayers(players)
  }

  const lockPawns = (playerId: string, pawns: Record<GameT.PawnSize, number>) => {
    if (playersStore.lockPawns(playerId, pawns)) {
      state.status = 'in-progress'
    }
  }

  const getLastTurn = () => {
    return Math.max(...Object.keys(state.turns).map(Number), 0)
  }

  const getCurrentTurn = () => {
    return getLastTurn() + 1
  }

  const manageAllowedPawns = () => {
    const lastTurn = getLastTurn()
    const rejectConditions = [
      lastTurn < GameService.UNLOCK_MEDIUM_PAWNS_TURN,
      state.allowedPawns.length === 2 && lastTurn < GameService.UNLOCK_BIG_PAWNS_TURN,
      lastTurn > GameService.UNLOCK_BIG_PAWNS_TURN,
    ]

    if (rejectConditions.some(Boolean)) return
    if (lastTurn === GameService.UNLOCK_MEDIUM_PAWNS_TURN) state.allowedPawns.push('medium')
    if (lastTurn === GameService.UNLOCK_BIG_PAWNS_TURN) state.allowedPawns.push('big')
  }

  const nextTurn = (payload?: GameT.PutPawnPayload | GameT.ApplyShieldPayload) => {
    playersStore.decreaseParalyzeProtection(state.currentPlayerId)
    state.turns[getLastTurn() + 1] = { playerId: state.currentPlayerId, move: payload || 'skip' }
    state.currentPlayerId = playersStore
      .getPlayers()
      .find((player) => player !== state.currentPlayerId)!

    manageAllowedPawns()
  }

  const checkInstantWin = () => {
    const CAPTURED_PAWNS_FOR_INSTANT_WIN = 5
    const player = playersStore.getPlayer(state.currentPlayerId)

    return player.capturedPawnsCounter === CAPTURED_PAWNS_FOR_INSTANT_WIN
  }

  const isLineCaptureWin = (): string | undefined => {
    const sequences = GameService.findWinningLine(state.board)

    if (state.sequences.length > 0) {
      const sequence = state.sequences.find((cell) => sequences.includes(cell))

      if (sequence) {
        const [playerId] = sequence.split('/')
        return playerId
      }
    }

    state.sequences = sequences

    return undefined
  }

  const skipTurn = () => {
    const result = playersStore.updateSkippedTurnsCount(state.currentPlayerId)
    const opponentId = playersStore.getOpponentId(state.currentPlayerId)

    const penalties: Record<number, () => void> = {
      1: () => playersStore.manipulatePawnAmount(opponentId, 'small', 1),
      2: () => playersStore.manipulatePawnAmount(opponentId, 'medium', 1),
      3: () => playersStore.manipulatePawnAmount(opponentId, 'big', 1),
      4: () => playersStore.addSpecialMove(opponentId, 'shield'),
      5: () => playersStore.addSpecialMove(opponentId, 'drop'),
    }

    penalties[result]()

    playersStore.removeParalyzeEffect(state.currentPlayerId)

    nextTurn()
  }

  const checkWinner = (payload: GameT.MovePayload) => {
    const isInstantWin = checkInstantWin()
    state.winner = isInstantWin ? state.currentPlayerId : isLineCaptureWin()

    if (!isInstantWin) nextTurn(payload)
    if (state.currentPlayerId === state.winner) {
      state.status = 'finished'
    }
  }

  const isCellProtected = ({ rowIndex, columnIndex }: GameT.Location) => {
    return state.board[rowIndex][columnIndex].shield.activeInTurn === getCurrentTurn()
  }

  const isPutPawnAllowed = (payload: GameT.PutPawnPayload) => {
    const player = playersStore.getPlayer(state.currentPlayerId)
    const { rowIndex, columnIndex, pawnSize } = payload
    const isPawnAvailable = player.pawns[payload.pawnSize] > 0
    const isPawnSizeAllowed = state.allowedPawns.includes(pawnSize)

    return (
      isPawnAvailable &&
      isPawnSizeAllowed &&
      !isCellProtected({ rowIndex, columnIndex }) &&
      !playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive
    )
  }

  const isDropPawnAllowed = (payload: GameT.DropOpponentPawnPayload) => {
    const { rowIndex, columnIndex } = payload
    const isPlayerPawn = state.board[rowIndex][columnIndex].pawn?.playerId === state.currentPlayerId

    return (
      !isCellProtected({ rowIndex, columnIndex }) &&
      !isPlayerPawn &&
      !playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive
    )
  }

  const dropOpponentPawn = (payload: GameT.MovePayload) => {
    if (!isDropPawnAllowed(payload)) return
    if (playersStore.useSpecialMove(state.currentPlayerId, 'drop') === 'not-allowed') return

    const { rowIndex, columnIndex } = payload
    state.board[rowIndex][columnIndex].pawn = null

    checkWinner(payload)
  }

  const paralyzePlayer = () => {
    const opponentId = playersStore.getOpponentId(state.currentPlayerId)
    const result = SpecialEffectsSequences.paralyze.every(([row, column]) => {
      return state.board[row][column].pawn?.playerId === state.currentPlayerId
    })

    if (!result || !playersStore.canPlayerBeParalyzed(opponentId)) return

    playersStore.paralyzePlayer(playersStore.getOpponentId(state.currentPlayerId))
  }

  const putPawn = (payload: GameT.PutPawnPayload) => {
    const { rowIndex, columnIndex, pawnSize } = payload
    const cellPawn = state.board[rowIndex][columnIndex].pawn

    if (!isPutPawnAllowed(payload) || state.status !== 'in-progress') return

    if (cellPawn) {
      if (!GameService.isPawnBigger(pawnSize, cellPawn.size)) return
      if (cellPawn.playerId !== state.currentPlayerId) {
        playersStore.updatePlayerCapturedPawns(state.currentPlayerId)
      }
    }

    state.board[rowIndex][columnIndex].pawn = {
      size: pawnSize,
      playerId: state.currentPlayerId,
    }
    playersStore.manipulatePawnAmount(state.currentPlayerId, pawnSize, -1)

    paralyzePlayer()
    checkWinner(payload)
  }

  const applyShield = (payload: GameT.ApplyShieldPayload) => {
    if (playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive) return
    if (playersStore.useSpecialMove(state.currentPlayerId, 'shield') === 'not-allowed') return
    const { rowIndex, columnIndex } = payload

    state.board[rowIndex][columnIndex].shield.appliedBy = state.currentPlayerId
    state.board[rowIndex][columnIndex].shield.activeInTurn = getLastTurn() + 2

    checkWinner(payload)
  }

  return {
    state,
    initiateGame,
    putPawn,
    lockPawns,
    applyShield,
    getCurrentTurn,
    isLineCaptureWin,
    dropOpponentPawn,
    skipTurn,
  }
})
