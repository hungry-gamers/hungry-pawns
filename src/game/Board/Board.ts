export class Board {
  static size = '3x3'

  constructor() {}

  calculateCells() {
    const [rows, columns] = Board.size.split('x')

    return +rows * +columns
  }
}
