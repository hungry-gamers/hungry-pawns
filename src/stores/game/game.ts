import { reactive } from 'vue'
import { defineStore } from 'pinia'
import type { Game, PutPawnPayload, Player, PawnSize } from '@/stores/game/types.ts'

export const useGameStore = defineStore('game', () => {
  const state = reactive<Game>({
    board: Array.from({ length: 3 }, () => Array(3).fill(null)),
    currentPlayerId: '',
    pawns: {},
  })

  const initiateGame = (players: Player[]) => {
    state.board = Array.from({ length: 3 }, () => Array(3).fill(null))
    state.currentPlayerId = players[0].id

    players.forEach((player) => {
      state.pawns[player.id] = { ...player.pawns }
    })
  }

  const isBigger = (size: PawnSize, pawnOnBoardSize: PawnSize) => {
    const order: PawnSize[] = ['small', 'medium', 'big']
    return order.indexOf(size) > order.indexOf(pawnOnBoardSize)
  }

  const putPawn = (payload: PutPawnPayload) => {
    if (!state.pawns[state.currentPlayerId][payload.pawn.size]) return
    const cellPawn = state.board[payload.rowIndex][payload.columnIndex]

    if (!cellPawn || isBigger(payload.pawn.size, cellPawn.size)) {
      const players = Object.keys(state.pawns)
      state.board[payload.rowIndex][payload.columnIndex] = payload.pawn
      state.pawns[state.currentPlayerId][payload.pawn.size]--
      state.currentPlayerId = players.find((player) => player !== state.currentPlayerId)
    }
  }

  return {
    state,
    initiateGame,
    putPawn,
  }
})
