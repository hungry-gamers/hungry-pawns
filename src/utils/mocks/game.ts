import type { Player } from '@/stores/game/types.ts'

export const players: Player[] = [
  { id: '1', pawns: { small: 3, medium: 3, big: 3 } },
  { id: '2', pawns: { small: 3, medium: 3, big: 3 } },
]
