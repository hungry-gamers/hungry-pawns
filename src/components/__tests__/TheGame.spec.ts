import { describe, it, expect, vi } from 'vitest'
import { createTestingPinia } from '@pinia/testing'

import { mount } from '@vue/test-utils'
import Board from '../TheGame.vue'
import { useGameStore } from '../../stores/game/game'

describe('Game', () => {
  const wrapper = mount(Board, {
    props: { pawnSize: 'small' },
    global: { plugins: [createTestingPinia({ createSpy: vi.fn })] },
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should put pawn on the board', () => {
    const gameStore = useGameStore()
    const cell = wrapper.find('[data-test-id="cell-0-0"]')
    gameStore.state.currentPlayerId = '1'

    cell.trigger('click')

    expect(gameStore.putPawn).toHaveBeenCalledTimes(1)
    expect(gameStore.putPawn).toHaveBeenCalledWith({
      pawn: { size: 'small', player: '1' },
      rowIndex: 0,
      columnIndex: 0,
    })
  })
})
