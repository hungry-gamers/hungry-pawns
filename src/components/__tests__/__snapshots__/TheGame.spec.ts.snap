// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game > should render correctly 1`] = `
"<div data-v-1a14ab8b="" id="board">
  <div data-v-1a14ab8b="" class="row">
    <div data-v-1a14ab8b="" data-test-id="cell-0-0" class="cell">
      <!--v-if-->
    </div>
    <div data-v-1a14ab8b="" data-test-id="cell-0-1" class="cell">
      <!--v-if-->
    </div>
    <div data-v-1a14ab8b="" data-test-id="cell-0-2" class="cell">
      <!--v-if-->
    </div>
  </div>
  <div data-v-1a14ab8b="" class="row">
    <div data-v-1a14ab8b="" data-test-id="cell-1-0" class="cell">
      <!--v-if-->
    </div>
    <div data-v-1a14ab8b="" data-test-id="cell-1-1" class="cell">
      <!--v-if-->
    </div>
    <div data-v-1a14ab8b="" data-test-id="cell-1-2" class="cell">
      <!--v-if-->
    </div>
  </div>
  <div data-v-1a14ab8b="" class="row">
    <div data-v-1a14ab8b="" data-test-id="cell-2-0" class="cell">
      <!--v-if-->
    </div>
    <div data-v-1a14ab8b="" data-test-id="cell-2-1" class="cell">
      <!--v-if-->
    </div>
    <div data-v-1a14ab8b="" data-test-id="cell-2-2" class="cell">
      <!--v-if-->
    </div>
  </div>
</div>"
`;
